File: Leetcode/src/com/leetcode/_100_SameTree.java
package com.leetcode;

public class _100_SameTree {

    public class TreeNode{
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) {
            val = x;
        }
    }
    /**
     * 100. Same Tree
     * When: 2019/04/15
     *
     * solution:
     * 递归两棵树的左边全部和右边全部
     * @param p
     * @param q
     * @return
     */
    public boolean isSameTree(TreeNode p, TreeNode q) {
       if (p == null && q == null) return true;
       if (p == null || q == null) return false;
       if (p.val != q.val) return false;
       return isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
    }
}

File: Leetcode/src/com/leetcode/tree/preorder/_100_same_tree.py
from collections import deque
from typing import Optional

from leetcode.common.py_utils import TreeNode


class Solution:
    # BFS
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        queue = deque()
        queue.append(p)
        queue.append(q)
        while queue:
            first = queue.popleft()
            second = queue.popleft()
            if first and second:
                if first.val != second.val:
                    return False
            elif first or second:
                return False
            else:
                # first and second both are None
                continue
            queue.append(first.left)
            queue.append(second.left)
            queue.append(first.right)
            queue.append(second.right)
        return True


class Solution2:
    # DFS
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if not p and not q:
            return True
        if not p or not q:
            return False

        # p and q is not None
        if p.val != q.val:
            return False

        return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)
